altosui/altosui.1
.\"
.\" Copyright © 2010 Bdale Garbee <bdale@gag.com>
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
.\"
.\"
.TH ALTOSUI 1 "altosui" ""
.SH NAME
altosui \- Rocket flight monitor
.SH SYNOPSIS
.B "altosui"
.SH DESCRIPTION
.I altosui
connects to a TeleDongle or TeleMetrum device through a USB serial device.
It provides a menu-oriented
user interface to monitor, record and review rocket flight data.
.SH USAGE
When connected to a TeleDongle device, altosui turns on the radio
receiver and listens for telemetry packets. It displays the received
telemetry data, and reports flight status via voice synthesis. All
received telemetry information is recorded to a file.
.P
When connected to a TeleMetrum device, altosui can be used to configure the
TeleMetrum, and to downloads the eeprom data and store it in a file.
.P
A number of other menu options exist, including the ability to export flight
data in different formats.
.SH FILES
All data log files are recorded into a user-specified directory
(default ~/TeleMetrum). Files are named using the current date, the serial
number of the reporting device, the flight number recorded in the data
and either '.telem' for telemetry data or '.eeprom' for eeprom data.
.SH AUTHOR
Keith Packard


ChangeLog
commit ea95c060f8bd959cefc60dd6d411151f3c91384e
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 1 00:12:05 2010 -0700

    fix symlink paths in rules file

commit f95220b6f0cc74928e0e7d7c7e1f443eafc32a02
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 1 00:03:23 2010 -0700

    update changelogs for Debian build

commit ea7130e953622884afc348265f5a4c58f9876823
Author: Bdale Garbee <bdale@gag.com>
Date:   Wed Dec 1 00:00:58 2010 -0700

    modify Debian package build to deliver one copy of background.png for
    the themes and symlink it into place for gdm and slim

commit 3696b30a8d1eb351353d84b5f7af8eeff55c468b
Author: Bdale Garbee <bdale@gag.com>
Date:   Mon Nov 29 21:47:13 2010 -0700

    update changelogs for Debian build

commit 11b3f9ff715017a2dec02003275885334f22c009
Author: Bdale Garbee <bdale@gag.com>
Date:   Mon Nov 29 21:46:02 2010 -0700

    releasing 0.8.1

commit 5f3f11e11dff9b9f4d1dde279c0d474de0de12a5
Author: Keith Packard <keithp@keithp.com>
Date:   Mon Nov 29 20:41:49 2010 -0800

    Add minimal release testing plan to Releasing
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit f0a4deee23984a8f779917bbeaf74a66a0abf592
Author: Keith Packard <keithp@keithp.com>
Date:   Mon Nov 29 20:34:27 2010 -0800

    Move "Releasing" to top level
    
    This file documents the AltOS release process for all systems.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit b727156ef0a7fb6e442ca28be27eb344a213ecf8
Author: Keith Packard <keithp@keithp.com>
Date:   Mon Nov 29 20:17:35 2010 -0800

    windows: Add jfreechart.jar and jcommon.jar to windows install image
    
    These are necessary for the Graph Data button to do anything useful.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 3e2220a180f95971d222a597d2057ca328c27356
Author: Bdale Garbee <bdale@gag.com>
Date:   Mon Nov 29 21:29:14 2010 -0700

    adding Bdale's release process document to the source tree

commit 61a924099800494b589cbbb87c65b552ccbd8394
Author: Bdale Garbee <bdale@gag.com>
Date:   Mon Nov 29 14:40:27 2010 -0700

    fix an Altos vs Altus typo in the docs

commit e840b6594b8a939f148fa7231e1b06a280d94074
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Nov 28 22:42:43 2010 -0700

    fix section layering

commit f39698bbc12afdfadfac56c90030e16db93cf4fc
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Nov 28 19:46:03 2010 -0700

    fix publish target in doc/Makefile

commit 13cea7a96821165a10a8b2433af1da7508882b0a
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Nov 28 18:48:31 2010 -0700

    moved doc dir in web content to AltOS tree

commit 6f3c72462af86a7d9dec20f4c813a524a5be9fbb
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Nov 28 18:32:17 2010 -0700

    update changelogs for Debian build

commit a375942979dbcd8239d8c0addb10616e6048f6ea
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Nov 28 18:31:26 2010 -0700

    rewind changelog to recover from stupid build failure

commit 263cf6c4fc5c1a240d719bb0ceb33393864fb3d6
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Nov 28 18:24:00 2010 -0700

    update changelogs for Debian build

commit 2615289265b6f8fa08827be794b4eee569fc6333
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Nov 28 18:23:17 2010 -0700

    update changelogs for Debian build

commit 5e23bcc2d2ad33a839fedd2158213098a05cd9ae
Author: Bdale Garbee <bdale@gag.com>
Date:   Sun Nov 28 18:21:31 2010 -0700

    declaring 0.8 released

commit b09d3e11c32abd07f8e11b483ac4b57148f4f2d8
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Nov 27 23:08:41 2010 -0800

    windows: Bump .inf file version

commit 42c2394ccc1a1ee6dc134ed963a4f79acf031d0b
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Nov 27 23:04:50 2010 -0800

    windows: more .inf file hacking
    
    Add a LayoutFile reference
    Add the FakeModemCopyFileSection
    
    Copy AltosMetrum.Install section to AltusMetrum.Install.NT section. I
    don't know what this might do; existing files seem inconsistent.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 26369a92398f4ebe3fcd54ce34cebffae0cfdf07
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Nov 27 22:39:31 2010 -0800

    windows: remove some non-existent .inf file section references
    
    The mfglt and VerboseResultCodes sections are not present in the .inf
    file, so remove refernces to them.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit 641c5373724d34c3adfcf42420a528d6bba736b9
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Nov 27 22:35:18 2010 -0800

    windows: try harder to get windows install to work
    
    Add devIDs to .nsi file. Fix install section name mapping from the
    hot-plug info.
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit cd414e2c04ce5ecbc75f19325a6d6f82cd489fb3
Author: Keith Packard <keithp@keithp.com>
Date:   Sat Nov 27 16:30:29 2010 -0800

    altosui: Correct windows hardware IDs for nsis installer file
    
    Need real hardware IDs (encoded USB ids) to get windows to
    auto-install the driver?
    
    Signed-off-by: Keith Packard <keithp@keithp.com>

commit f834b22905db87557f729f942607dc363b013694
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Nov 25 22:31:33 2010 -0700

    update changelogs for Debian build

commit da3b39d6b1b5ba48686c2a4add4b5448fd5711af
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Nov 25 22:29:01 2010 -0700

    changed main document name from telemetrum-doc to altusmetrum

commit ab6c9c983a2830bc8807e1b75d2576141b73632d
Author: Bdale Garbee <bdale@gag.com>
Date:   Thu Nov 25 19:30:02 2010 -0700

    update fat target in Makefiles to reflect move of altosui and libaltos

commit b8f05cdc0e9b4a96852eed9d38ff6d5950e2d2ed
Author: Keith Packard <keithp@keithp.com>
Date:   Thu Nov 25 17:29:28 2010 -0800

    altosui: Clean up flash code to ensure swing gets called from right thread
    
    This moves all of the flash code to a separate thread and passes

altosui/AltosUI.java
/*
 * Copyright © 2010 Keith Packard <keithp@keithp.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA.
 */

package altosui;

import java.awt.*;
import java.awt.event.*;
import javax.swing.*;
import javax.swing.filechooser.FileNameExtensionFilter;
import javax.swing.table.*;
import java.io.*;
import java.util.*;
import java.text.*;
import java.util.prefs.*;
import java.util.concurrent.LinkedBlockingQueue;

import libaltosJNI.*;

public class AltosUI extends JFrame {
	public AltosVoice voice = new AltosVoice();

	public static boolean load_library(Frame frame) {
		if (!AltosDevice.load_library()) {
			JOptionPane.showMessageDialog(frame,
						      String.format("No AltOS library in \"%s\"",
								    System.getProperty("java.library.path","<undefined>")),
						      "Cannot load device access library",
						      JOptionPane.ERROR_MESSAGE);
			return false;
		}
		return true;
	}

	void telemetry_window(AltosDevice device) {
		try {
			AltosFlightReader reader = new AltosTelemetryReader(device);
			if (reader != null)
				new AltosFlightUI(voice, reader, device.getSerial());
		} catch (FileNotFoundException ee) {
			JOptionPane.showMessageDialog(AltosUI.this,
						      String.format("Cannot open device \"%s\"",
								    device.toShortString()),
						      "Cannot open target device",
						      JOptionPane.ERROR_MESSAGE);
		} catch (AltosSerialInUseException si) {
			JOptionPane.showMessageDialog(AltosUI.this,
						      String.format("Device \"%s\" already in use",
								    device.toShortString()),
						      "Device in use",
						      JOptionPane.ERROR_MESSAGE);
		} catch (IOException ee) {
			JOptionPane.showMessageDialog(AltosUI.this,
						      device.toShortString(),
						      "Unkonwn I/O error",
						      JOptionPane.ERROR_MESSAGE);
		}
	}

	Container	pane;
	GridBagLayout	gridbag;

	JButton addButton(int x, int y, String label) {
		GridBagConstraints	c;
		JButton			b;

		c = new GridBagConstraints();
		c.gridx = x; c.gridy = y;
		c.fill = GridBagConstraints.BOTH;
		c.weightx = 1;
		c.weighty = 1;
		b = new JButton(label);

		Dimension ps = b.getPreferredSize();

		gridbag.setConstraints(b, c);
		add(b, c);
		return b;
	}

	public AltosUI() {

		load_library(null);

		java.net.URL imgURL = AltosUI.class.getResource("/altus-metrum-16x16.jpg");
		if (imgURL != null)
			setIconImage(new ImageIcon(imgURL).getImage());

		AltosPreferences.init(this);

		pane = getContentPane();
		gridbag = new GridBagLayout();
		pane.setLayout(gridbag);

		JButton	b;

		b = addButton(0, 0, "Monitor Flight");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						ConnectToDevice();
					}
				});
		b = addButton(1, 0, "Save Flight Data");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						SaveFlightData();
					}
				});
		b = addButton(2, 0, "Replay Flight");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						Replay();
					}
				});
		b = addButton(3, 0, "Graph Data");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						GraphData();
					}
				});
		b = addButton(4, 0, "Export Data");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						ExportData();
					}
				});
		b = addButton(0, 1, "Configure TeleMetrum");
		b.addActionListener(new ActionListener() {
					public void actionPerformed(ActionEvent e) {
						ConfigureTeleMetrum();
					}
				});

		b = addButton(1, 1, "Configure AltosUI");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					ConfigureAltosUI();
				}
			});

		b = addButton(2, 1, "Flash Image");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					FlashImage();
				}
			});

		b = addButton(3, 1, "Fire Igniter");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					FireIgniter();
				}
			});

		b = addButton(4, 1, "Quit");
		b.addActionListener(new ActionListener() {
				public void actionPerformed(ActionEvent e) {
					System.exit(0);
				}
			});

		setTitle("AltOS");

		pane.doLayout();
		pane.validate();

		doLayout();
		validate();

		setVisible(true);

		Insets i = getInsets();
		Dimension ps = rootPane.getPreferredSize();
		ps.width += i.left + i.right;
		ps.height += i.top + i.bottom;
		setPreferredSize(ps);
		setSize(ps);
		setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
		addWindowListener(new WindowAdapter() {
			@Override
			public void windowClosing(WindowEvent e) {
				System.exit(0);
			}
		});
	}

	private void ConnectToDevice() {
		AltosDevice	device = AltosDeviceDialog.show(AltosUI.this,
								AltosDevice.product_basestation);

		if (device != null)
			telemetry_window(device);
	}

	void ConfigureCallsign() {
		String	result;
		result = JOptionPane.showInputDialog(AltosUI.this,
						     "Configure Callsign",
						     AltosPreferences.callsign());
		if (result != null)
			AltosPreferences.set_callsign(result);
	}

	void ConfigureTeleMetrum() {
		new AltosConfig(AltosUI.this);
	}

	void FlashImage() {
		new AltosFlashUI(AltosUI.this);
	}

	void FireIgniter() {
		new AltosIgniteUI(AltosUI.this);
	}

	/*
	 * Replay a flight from telemetry data
	 */
	private void Replay() {
		AltosDataChooser chooser = new AltosDataChooser(
			AltosUI.this);

		AltosRecordIterable iterable = chooser.runDialog();
		if (iterable != null) {
			AltosFlightReader reader = new AltosReplayReader(iterable.iterator(),
									 chooser.filename());
			new AltosFlightUI(voice, reader);
		}
	}

	/* Connect to TeleMetrum, either directly or through
	 * a TeleDongle over the packet link
	 */
	private void SaveFlightData() {
		new AltosEepromDownload(AltosUI.this);
	}

	/* Load a flight log file and write out a CSV file containing
	 * all of the data in standard units
	 */

	private void ExportData() {
		AltosDataChooser chooser;
		chooser = new AltosDataChooser(this);
		AltosRecordIterable record_reader = chooser.runDialog();
		if (record_reader == null)
			return;
		new AltosCSVUI(AltosUI.this, record_reader, chooser.file());
	}

	/* Load a flight log CSV file and display a pretty graph.
	 */

	private void GraphData() {
		AltosDataChooser chooser;
		chooser = new AltosDataChooser(this);
		AltosRecordIterable record_reader = chooser.runDialog();
		if (record_reader == null)
			return;
		new AltosGraphUI(record_reader);
	}

	private void ConfigureAltosUI() {
		new AltosConfigureUI(AltosUI.this, voice);
	}

	static AltosRecordIterable open_logfile(String filename) {
		File file = new File (filename);
		try {
			FileInputStream in;

			in = new FileInputStream(file);
			if (filename.endsWith("eeprom"))
				return new AltosEepromIterable(in);
			else
				return new AltosTelemetryIterable(in);
		} catch (FileNotFoundException fe) {
			System.out.printf("Cannot open '%s'\n", filename);
			return null;
		}
	}

	static AltosWriter open_csv(String filename) {
		File file = new File (filename);
		try {
			return new AltosCSV(file);
		} catch (FileNotFoundException fe) {
			System.out.printf("Cannot open '%s'\n", filename);
			return null;
		}
	}

	static AltosWriter open_kml(String filename) {
		File file = new File (filename);
		try {
			return new AltosKML(file);
		} catch (FileNotFoundException fe) {
			System.out.printf("Cannot open '%s'\n", filename);
			return null;
		}
	}

	static final int process_csv = 1;
	static final int process_kml = 2;

	static void process_file(String input, int process) {
		AltosRecordIterable iterable = open_logfile(input);
		if (iterable == null)
			return;
		if (process == 0)
			process = process_csv;
		if ((process & process_csv) != 0) {
			String output = Altos.replace_extension(input,".csv");
			System.out.printf("Processing \"%s\" to \"%s\"\n", input, output);
			if (input.equals(output)) {
				System.out.printf("Not processing '%s'\n", input);
			} else {
				AltosWriter writer = open_csv(output);
				if (writer != null) {
					writer.write(iterable);
					writer.close();
				}
			}
		}
		if ((process & process_kml) != 0) {
			String output = Altos.replace_extension(input,".kml");
			System.out.printf("Processing \"%s\" to \"%s\"\n", input, output);
			if (input.equals(output)) {
				System.out.printf("Not processing '%s'\n", input);
			} else {
				AltosWriter writer = open_kml(output);
				if (writer == null)
					return;
				writer.write(iterable);
				writer.close();
			}
		}
	}

	public static void main(final String[] args) {
		int	process = 0;
		/* Handle batch-mode */
        if (args.length == 1 && args[0].equals("--help")) {
		System.out.printf("Usage: altosui [OPTION]... [FILE]...\n");
		System.out.printf("  Options:\n");
		System.out.printf("    --fetchmaps <lat> <lon>\tpre-fetch maps for site map view\n");
		System.out.printf("    --replay <filename>\t\trelive the glory of past flights \n");
		System.out.printf("    --csv\tgenerate comma separated output for spreadsheets, etc\n");
		System.out.printf("    --kml\tgenerate KML output for use with Google Earth\n");
        } else if (args.length == 3 && args[0].equals("--fetchmaps")) {
            double lat = Double.parseDouble(args[1]);
            double lon = Double.parseDouble(args[2]);
            AltosSiteMap.prefetchMaps(lat, lon, 5, 5);
        } else if (args.length == 2 && args[0].equals("--replay")) {
			String filename = args[1];
			FileInputStream in;
			try {
				in = new FileInputStream(filename);
			} catch (Exception e) {
				System.out.printf("Failed to open file '%s'\n", filename);
				return;
			}
			AltosRecordIterable recs;
			AltosReplayReader reader;
			if (filename.endsWith("eeprom")) {
				recs = new AltosEepromIterable(in);
			} else {
				recs = new AltosTelemetryIterable(in);
			}
			reader = new AltosReplayReader(recs.iterator(), filename);
			AltosFlightUI flight_ui = new AltosFlightUI(new AltosVoice(), reader);
			flight_ui.set_exit_on_close();
			return;
		} else if (args.length > 0) {
			for (int i = 0; i < args.length; i++) {
				if (args[i].equals("--kml"))
					process |= process_kml;
				else if (args[i].equals("--csv"))
					process |= process_csv;
				else
					process_file(args[i], process);
			}
		} else {
			AltosUI altosui = new AltosUI();
			altosui.setVisible(true);

			AltosDevice[] devices = AltosDevice.list(AltosDevice.product_basestation);
			for (int i = 0; i < devices.length; i++)
				altosui.telemetry_window(devices[i]);
		}
	}
}
